### PR TITLE
Fully revise cloudwatch support

### PR DESCRIPTION
**Pull request by zakwilson**

### CODE REVIEW

File: services/cloud_watch.rb [hunk 1]
```diff
@@ -1,126 +1,91 @@
...
-  def aws_connection
-    @aws_connection ||= begin
-      http.dup.tap do |c|
-        c.builder.insert(0, AwsAuthentication, settings[:aws_access_key_id], settings[:aws_secret_access_key])
-        c.builder.insert(1, Faraday::Request::UrlEncoded)
-        c.builder.insert(3, Faraday::Response::RaiseError)
-      end
-    end
-  end
+  def prepare_post_data(events, size_limit=8192, max_days = 14)
```
Comment (troy) [inline services/cloud_watch.rb]: In as much as there's a coding style in this repo, it's `k = v` :)

File: services/cloud_watch.rb [hunk 2]
```diff
@@ -1,126 +1,91 @@
...
-
-      match_data = event[:message].match(regex)
-      # does not match event
-      next unless match_data
-
-      # use extracted value or 1 (no regex backreference provided)
-      value = match_data[1] ? match_data[1].to_i : 1
+      timestamp = Time.iso8601(event[:received_at]).to_i
+      counts[timestamp] += 1
+    end
```
Comment (troy) [inline services/cloud_watch.rb]: Can you break these ~7 lines into a new helper method, like `counts = event_counts_by_received_at(events)`? `Stathat` can use it too.

And add unit test coverage for that method?

File: services/cloud_watch.rb [hunk 3]
```diff
@@ -1,126 +1,91 @@
...
 
-      data_points << {
-        :metric_name => name,
-        :value => value,
-        :timestamp => event[:received_at]
+    metric_data = counts.map do |time, count|
+      timestamp = Time.at(time)
+      if timestamp < Time.now - 60*60*24*14
+        raise_config_error "Cloudwatch will not accept #{timestamp.iso8601} because it is more than #{max_days} days old"
```
Comment (troy) [inline services/cloud_watch.rb]: Nice catch with identifying this case

File: services/cloud_watch.rb [hunk 4]
```diff
@@ -1,126 +1,91 @@
...
+      end
+      {
+        metric_name: settings[:cloudwatch_metric_name],
```
Comment (troy) [inline services/cloud_watch.rb]: Papertrail's param names will generally omit the service name, so let's use the shorter `metric_name` and `namespace` names.

File: services/cloud_watch.rb [hunk 5]
```diff
@@ -1,126 +1,91 @@
...
+        timestamp: timestamp.iso8601,
+        statistic_values: { sample_count: count,
+                            sum: count,
+                            minimum: count,
+                            maximum: count,
+                          },
```
Comment (troy) [inline services/cloud_watch.rb]: http://docs.aws.amazon.com/AmazonCloudWatch/latest/DeveloperGuide/cloudwatch_concepts.html#Statistic makes me think we should just submit the `sum` here. Do you have an opinion? I can explore the CW docs more if you don't, or feel free to recommend behavior.

Comment (zakwilson) [inline services/cloud_watch.rb]: The other parameters are required per http://docs.aws.amazon.com/sdkforruby/api/Aws/CloudWatch/Types/MetricDatum.html

File: services/cloud_watch.rb (hunk 2 above)
Comment (zakwilson) [inline services/cloud_watch.rb]: Should the helper method go in a new file in lib/papertrail_services/helpers/, or in log_helpers.rb or...?

Comment (troy) [inline services/cloud_watch.rb]: Yeah, let's put it in `LogHelpers` alongside `syslog_format()`

File: services/cloud_watch.rb (hunk 5 above)
Comment (troy) [inline services/cloud_watch.rb]: That makes sense. Can we omit `statistic_values` entirely and just submit `value`? Something like [this](http://docs.aws.amazon.com/cli/latest/reference/cloudwatch/put-metric-data.html) (at the very end) is what I'm thinking of.

Comment (zakwilson) [inline services/cloud_watch.rb]: Yes, that actually looks like it works.

Comment (troy) [inline services/cloud_watch.rb]: Cool :)

File: test/cloud_watch_test.rb [hunk 6]
```diff
@@ -1,28 +1,40 @@
...
+      @svc.prepare_post_data(@svc.payload[:events], size_limit=8)
+    }
+  end
 
-    svc.aws_connection.build do |b|
-      b.adapter :test do |stub|
-        stub.post('/') { [ 200, nil, '' ]}
-      end
-    end
+  def test_counts
```
Comment (troy) [inline test/cloud_watch_test.rb]: Nice!

**Comment by troy** [comment]
Assigning to me to merge and deploy.

**Pull request merged**